File: otto-mobile/otto-mobile/DriveDisplayNaming.swift
import Foundation

/// Display titles for drives — drive names are independent of route names.
enum DriveDisplayNaming {
    private static func isGenericDisplayName(_ name: String) -> Bool {
        let normalized = name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return normalized.isEmpty || normalized == "route drive" || normalized == "route drive session"
    }

    /// Default when starting a route drive (e.g. "Ocean Pass" → "Ocean Pass Drive").
    static func defaultTitle(fromRouteName routeName: String?) -> String {
        let route = routeName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if route.isEmpty || isGenericDisplayName(route) { return "Route Drive" }
        if route.lowercased().hasSuffix(" drive") { return route }
        return "\(route) Drive"
    }

    static func listTitle(routeName: String?, driveTitle: String?) -> String {
        let snapshotRoute = routeName?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        let raw = driveTitle?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if !raw.isEmpty, !isGenericDisplayName(raw) {
            if !snapshotRoute.isEmpty,
               raw == snapshotRoute,
               !raw.lowercased().hasSuffix(" drive") {
                return defaultTitle(fromRouteName: snapshotRoute)
            }
            return raw
        }
        return defaultTitle(fromRouteName: snapshotRoute.isEmpty ? routeName : snapshotRoute)
    }
}

File: otto-mobile/otto-android-LEGACY-DO-NOT-OPEN/app/src/main/java/to/ottomot/driftd/DriveDisplayNaming.kt
package to.ottomot.driftd

import to.ottomot.driftd.core.network.dto.DriveDto

/** Display titles for drives — drive names are independent of route names (iOS `DriveDisplayNaming`). */
object DriveDisplayNaming {
    private fun isGenericDisplayName(name: String): Boolean {
        val normalized = name.trim().lowercase()
        return normalized.isEmpty() || normalized == "route drive" || normalized == "route drive session"
    }

    fun defaultTitle(fromRouteName: String?): String {
        val route = fromRouteName?.trim().orEmpty()
        if (route.isEmpty() || isGenericDisplayName(route)) return "Route Drive"
        if (route.lowercase().endsWith(" drive")) return route
        return "$route Drive"
    }

    fun listTitle(routeName: String?, driveTitle: String?): String {
        val snapshotRoute = routeName?.trim().orEmpty()
        val raw = driveTitle?.trim().orEmpty()
        if (raw.isNotEmpty() && !isGenericDisplayName(raw)) {
            if (
                snapshotRoute.isNotEmpty() &&
                raw == snapshotRoute &&
                !raw.lowercase().endsWith(" drive")
            ) {
                return defaultTitle(snapshotRoute)
            }
            return raw
        }
        return defaultTitle(snapshotRoute.ifEmpty { routeName })
    }

    fun listTitle(drive: DriveDto): String =
        listTitle(routeName = drive.route?.name, driveTitle = drive.title)

    fun squadChatListTitle(driveTitle: String?): String =
        listTitle(routeName = null, driveTitle = driveTitle)
}
